profiling/service/service.go

```diff
@@ -0,0 +1,146 @@
+	// only the last few RPCs so that profiling stats do not grow unbounded. This
+	// parameter defines the upper limit on the number of RPCs for which
+	// statistics should be stored at any given time. An average RPC requires
+	// approximately 2-3 KiB of memory for profiling-related statistics, so
+	// choose an appropriate number based on the amount of memory you can afford.
+	StreamStatsSize uint32
+
+	// To expose the profiling service and its methods, a *grpc.Server must be
+	// provided.
+	Server *grpc.Server
+}
+
+var errorNilServer = errors.New("profiling: no grpc.Server provided")
+
+// Init takes a *ProfilingConfig to initialize profiling (turned on/off
+// depending on the value set in pc.Enabled) and register the profiling service
+// in the server provided in pc.Server.
+func Init(pc *ProfilingConfig) error {
+	if pc.Server == nil {
+		return errorNilServer
+	}
+
+	if err := profiling.InitStats(pc.StreamStatsSize); err != nil {
+		return err
+	}
+
+	ppb.RegisterProfilingServer(pc.Server, getProfilingServerInstance())
+
+	// Do this last after everything has been initialized and allocated.
+	profiling.Enable(pc.Enabled)
+
+	return nil
+}
+
+type profilingServer struct {
+	drainMutex sync.Mutex
+}
+
+var profilingServerInstance *profilingServer
+var profilingServerOnce sync.Once
+
+// getProfilingServerInstance creates and returns a singleton instance of
+// profilingServer. Only one instance of profilingServer is created to use a
+// shared mutex across all profilingServer instances.
+func getProfilingServerInstance() *profilingServer {
+	profilingServerOnce.Do(func() {
+		profilingServerInstance = &profilingServer{}
+	})
+
+	return profilingServerInstance
+}
+
+func (s *profilingServer) Enable(ctx context.Context, req *ppb.EnableRequest) (*ppb.EnableResponse, error) {
+	if req.Enabled {
+		grpclog.Infof("profilingServer: Enable: enabling profiling")
+	} else {
+		grpclog.Infof("profilingServer: Enable: disabling profiling")
+	}
+	profiling.Enable(req.Enabled)
+
+	return &ppb.EnableResponse{}, nil
+}
+
+func timerToProtoTimer(timer *profiling.Timer) *ppb.Timer {
+	return &ppb.Timer{
+		Tags:      timer.Tags,
+		BeginSec:  timer.Begin.Unix(),
+		BeginNsec: int32(timer.Begin.Nanosecond()),
+		EndSec:    timer.End.Unix(),
+		EndNsec:   int32(timer.End.Nanosecond()),
+		GoId:      timer.GoID,
+	}
+}
+
+func statToProtoStat(stat *profiling.Stat) *ppb.Stat {
+	protoStat := &ppb.Stat{
+		Tags:     stat.Tags,
+		Timers:   make([]*ppb.Timer, 0, len(stat.Timers)),
+		Metadata: stat.Metadata,
+	}
+	for _, t := range stat.Timers {
+		protoStat.Timers = append(protoStat.Timers, timerToProtoTimer(t))
+	}
+	return protoStat
+}
+
+func (s *profilingServer) GetStreamStats(ctx context.Context, req *ppb.GetStreamStatsRequest) (*ppb.GetStreamStatsResponse, error) {
+	// Since the drain operation is destructive, only one client request should
+	// be served at a time.
+	grpclog.Infof("profilingServer: GetStreamStats: processing request")
```

Comment:
Actually.. If profiling has been turned on and we are collecting data, this is quite rare.  It should be fine to log without the V(2).